File: Java/binary_tree_level_order_traversal.java
import java.util.*;

class TreeNode {
  int val;
  TreeNode left;
  TreeNode right;

  TreeNode(int x) {
    val = x;
  }
};

class LevelOrderTraversal {
  public static List<List<Integer>> traverse(TreeNode root) {
    List<List<Integer>> result = new ArrayList<List<Integer>>();
    if (root == null) {
      return result;
    }

    Queue<TreeNode> queue = new LinkedList<>();
    queue.offer(root);

    while (!queue.isEmpty()) {
      int queue_size = queue.size();
      List<Integer> curr_level = new ArrayList<>();

      for (int i = 0; i < queue_size; i++) {
        TreeNode curr_node = queue.poll();
        curr_level.add(curr_node.val);

        if (curr_node.left != null) {
          queue.offer(curr_node.left);
        }

        if (curr_node.right != null) {
          queue.offer(curr_node.right);
        }
      }
      result.add(curr_level);
    }
    return result;
  }


  public static void main(String[] args) {
    TreeNode root = new TreeNode(12);
    root.left = new TreeNode(7);
    root.right = new TreeNode(1);
    root.left.left = new TreeNode(9);
    root.right.left = new TreeNode(10);
    root.right.right = new TreeNode(5);
    List<List<Integer>> result = LevelOrderTraversal.traverse(root);
    System.out.println("Level order traversal: " + result);
  }
}

File: C++/binary_tree_level_order_traversal.cpp
#include <iostream>
#include <vector>
#include <queue>

using namespace std; 

struct TreeNode {
  int val;
  TreeNode *left;
  TreeNode *right;

  TreeNode(): val(0), left(nullptr), right(nullptr) {}
  TreeNode(int x): val(x), left(nullptr), right(nullptr) {}
  TreeNode(int x, TreeNode *left, TreeNode *right): val(x), left(left), right(right) {}
};

class solution {
public:
  vector<vector<int>> bfs(TreeNode *root) {
    
    vector<vector<int>> result;

    if (!root) {
      return result;
    }

    queue<TreeNode*> q;
    q.push(root);

    while (!q.empty()) {
      int level_size = q.size();
      vector<int> level_nodes;
      for (int i = 0; i < level_size; i++) {

        TreeNode *curr_node = q.front();
        q.pop();
        level_nodes.push_back(curr_node->val);

        if (curr_node->left) {
          q.push(curr_node->left);
        }
        if (curr_node->right) {
          q.push(curr_node->right);
        }

      }
      result.push_back(level_nodes);
    }
    return result;
  }
};
